mesh/nurbs.cpp

```diff
@@ -2211,6 +2214,244 @@ void NURBSExtension::CheckBdrPatches()
    }
 }
 
+void NURBSExtension::CheckKVDirection(int p, Array <int> &kvdir)
+{
+   Array<int> patchvert, edges, orient, edgevert;
+
+   // Get Element Vertices
+   patchTopo->GetElementVertices(p, patchvert);
+
+   // Get Element Edges
+   patchTopo->GetElementEdges(p, edges, orient);
+
+   // Match Element vertices to edges (knot direction)
+   kvdir.SetSize(Dimension());
+   kvdir = 0;
+
+   // Compare the vertices of the patches with the vertices of the knotvectors of knot2dge
+   // Based on the match the orientation will be a 1 or a -1
+   // -1: direction is flipped
+   //  1: direction is not  flipped
+   for (int i = 0; i < edges.Size(); i++)
+   {
+      patchTopo->GetEdgeVertices(edges[i], edgevert);
+      // First side
+      if (edgevert[0] == patchvert[0]  && edgevert[1] == patchvert[1])
+      {
+         kvdir[0] = 1;
+      }
+
+      if (edgevert[0] == patchvert[1]  && edgevert[1] == patchvert[0])
+      {
+         kvdir[0] = -1;
+      }
+
+      // Second side
+      if (edgevert[0] == patchvert[1]  && edgevert[1] == patchvert[2])
+      {
+         kvdir[1] = 1;
+      }
+
+      if (edgevert[0] == patchvert[2]  && edgevert[1] == patchvert[1])
+      {
+         kvdir[1] = -1;
+      }
+   }
+
+   // Third direction: only for 3D
+   if (Dimension() == 3)
+   {
+      for (int i = 0; i < edges.Size(); i++)
+      {
+         patchTopo->GetEdgeVertices(edges[i], edgevert);
+
+         if (edgevert[0] == patchvert[0]  && edgevert[1] == patchvert[4])
+         {
+            kvdir[2] = 1;
+         }
+
+         if (edgevert[0] == patchvert[4]  && edgevert[1] == patchvert[0])
+         {
+            kvdir[2] = -1;
+         }
+      }
+   }
+
+   // Verify that all knotvectors have been given a direction.
+   MFEM_ASSERT(kvdir.Find(0) == -1, "Could not find direction of knotvector.");
+}
+
+void NURBSExtension::CreateComprehensiveKV()
+{
+   Array<int> edges, orient, kvdir;
+
+   Array<int> e(Dimension());
+   if (Dimension() == 2)
+   {
+      e[0] = 0;
+      e[1] = 1;
+   }
+   else if (Dimension() == 3)
+   {
+      e[0] = 0;
+      e[1] = 3;
+      e[2] = 8;
+   }
+
+   // Create Comprehensive set of knotvectors based on direction of the knotvector.
+   knotVectorsCompr.SetSize(GetNP()*Dimension());
+   for (int p = 0; p < GetNP(); p++)
+   {
+      CheckKVDirection(p, kvdir);
+      patchTopo->GetElementEdges(p, edges, orient);
+
+      for (int i = 0; i < Dimension(); i++)
+      {
+         knotVectorsCompr[Dimension()*p+i] = new KnotVector(*(KnotVec(edges[e[i]])));
+         if (kvdir[i] == -1) {knotVectorsCompr[Dimension()*p+i]->Flip();}
+      }
+   }
+
+   MFEM_ASSERT(ConsistentUniqueKVComprehensiveKV(), "Mismatch in KnotVectors");
```

Comment:
In general, I prefer `MFEM_VERIFY` when performance is not a concern. Since this function is called only once, this will be checked only once, and the cost is negligible.

Reply:
Good suggestion! The effect on the performance should indeed be very small. I have changed it to a `MFEM_VERIFY`